torch/csrc/autograd/c2isl_function.cpp

```diff
@@ -0,0 +1,321 @@
+      break;
+    case at::ScalarType::Int:
+      t.code = DLDataTypeCode::kInt;
+      t.bits = 32;
+      break;
+    case at::ScalarType::Long:
+      t.code = DLDataTypeCode::kInt;
+      t.bits = 64;
+      break;
+    case at::ScalarType::Short:
+      t.code = DLDataTypeCode::kInt;
+      t.bits = 16;
+      break;
+    case at::ScalarType::Half:
+      t.code = DLDataTypeCode::kFloat;
+      t.bits = 16;
+      break;
+    case at::ScalarType::NumOptions:
+      throw std::logic_error("NumOptions is not a valid ScalarType");
+  }
+  return t;
+}
+
+tvm::Type toTVMType(at::ScalarType s) {
+  switch (s) {
+    case at::ScalarType::Byte:    return tvm::UInt(8);
+    case at::ScalarType::Char:    return tvm::Int(8);
+    case at::ScalarType::Double:  return tvm::Float(64);
+    case at::ScalarType::Float:   return tvm::Float(32);
+    case at::ScalarType::Int:     return tvm::Int(32);
+    case at::ScalarType::Long:    return tvm::Int(64);
+    case at::ScalarType::Short:   return tvm::Int(16);
+    case at::ScalarType::Half:    return tvm::Float(16);
+    case at::ScalarType::NumOptions:
+      throw std::logic_error("NumOptions is not a valid ScalarType");
+  }
+}
+
+tvm::Type toTVMType(const DLDataType& t) {
+  switch (t.code) {
+    case DLDataTypeCode::kUInt:  return tvm::UInt (t.bits, t.lanes);
+    case DLDataTypeCode::kInt:   return tvm::Int  (t.bits, t.lanes);
+    case DLDataTypeCode::kFloat: return tvm::Float(t.bits, t.lanes);
+    default:
+      throw std::logic_error("Unknown type code " + std::to_string(t.code));
+  }
+}
+
+// Types never die
+at::Type& toATenType(at::Backend backend, const DLDataType& t) {
+  at::ScalarType st;
+  switch (t.code) {
+    case DLDataTypeCode::kUInt:
+      switch (t.bits) {
+        case 8:  st = at::ScalarType::Byte;   break;
+        default: throw std::logic_error("Unsupported kUInt bits " + std::to_string(t.bits));
+      }
+      break;
+    case DLDataTypeCode::kInt:
+      switch (t.bits) {
+        case 8:  st = at::ScalarType::Char;   break;
+        case 16: st = at::ScalarType::Short;  break;
+        case 32: st = at::ScalarType::Int;    break;
+        case 64: st = at::ScalarType::Long;   break;
+        default: throw std::logic_error("Unsupported kInt bits " + std::to_string(t.bits));
+      }
+      break;
+    case DLDataTypeCode::kFloat:
+      switch (t.bits) {
+        case 16: st = at::ScalarType::Half;   break;
+        case 32: st = at::ScalarType::Float;  break;
+        case 64: st = at::ScalarType::Double; break;
+        default: throw std::logic_error("Unsupported kFloat bits " + std::to_string(t.bits));
+      }
+      break;
+    default:
+      throw std::logic_error("Unsupported code " + std::to_string(t.code));
+  }
+  // TODO: This doesn't put the tensor on the correct device
+  return getType(backend, st);
+}
+
+at::Type& toATenType(const DLContext& ctx, const DLDataType& t) {
+  if (t.lanes != 1) throw std::logic_error("ATen does not support lanes != 1");
```

Comment:
It seems better to move this check to the overload above. This way it'd be called in both code paths + this one doesn't even use `t`